mm2src/coins/eth/eth_swap_v2.rs

```diff
@@ -137,14 +504,378 @@ impl EthCoin {
+                    Token::FixedBytes(args.maker_secret_hash.to_vec()),
+                    Token::Address(token_address), // should be zero address Address::default()
+                ])?
+            },
+            EthCoinType::Erc20 { .. } => function.encode_input(&[
+                decoded[0].clone(), // id from erc20TakerPayment
+                decoded[1].clone(), // amount from erc20TakerPayment
+                decoded[2].clone(), // dexFee from erc20TakerPayment
+                decoded[4].clone(), // receiver from erc20TakerPayment
+                Token::FixedBytes(args.taker_secret_hash.to_vec()),
+                Token::FixedBytes(args.maker_secret_hash.to_vec()),
+                Token::Address(token_address), // erc20 token address from EthCoinType::Erc20
+            ])?,
+            EthCoinType::Nft { .. } => {
+                return Err(PrepareTxDataError::Internal("EthCoinType must be ETH or ERC20".into()))
+            },
+        };
+        Ok(data)
+    }
+
+    async fn prepare_spend_taker_payment_data(
+        &self,
+        args: &GenTakerPaymentSpendArgs<'_, Self>,
+        secret: &[u8],
+        decoded: Vec<Token>,
+        token_address: Address,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        let function = TAKER_SWAP_V2.function("spendTakerPayment")?;
+        let taker_address = public_to_address(args.taker_pub);
+        let data = function.encode_input(&[
+            decoded[0].clone(),                 // id from takerPaymentApprove
+            decoded[1].clone(),                 // amount from takerPaymentApprove
+            decoded[2].clone(),                 // dexFee from takerPaymentApprove
+            Token::Address(taker_address),      // taker address
+            decoded[4].clone(),                 // takerSecretHash from ethTakerPayment
+            Token::FixedBytes(secret.to_vec()), // makerSecret
+            Token::Address(token_address),      // tokenAddress
+        ])?;
+        Ok(data)
+    }
+
+    /// Retrieves the payment status from a given smart contract address based on the swap ID and state type.
+    pub(crate) async fn payment_status_v2(
+        &self,
+        swap_address: Address,
+        swap_id: Token,
+        contract_abi: &Contract,
+        payment_type: EthPaymentType,
+        state_index: usize,
+    ) -> Result<U256, PaymentStatusErr> {
+        let function_name = payment_type.as_str();
+        let function = contract_abi.function(function_name)?;
+        let data = function.encode_input(&[swap_id])?;
+        let bytes = self
+            .call_request(self.my_addr().await, swap_address, None, Some(data.into()))
+            .await?;
+        let decoded_tokens = function.decode_output(&bytes.0)?;
+
+        let state = decoded_tokens.get(state_index).ok_or_else(|| {
+            PaymentStatusErr::Internal(ERRL!(
+                "Payment status must contain 'state' as the {} token",
+                state_index
+            ))
+        })?;
```

Comment:
not an Internal error... `InvalidData`

Reply:
it is supposed that input data to some function was decoded correctly, so the problem should be in wrong index which we try get. Its more about internal issues in logic or its a bug
```rust
        let data = function.encode_input(&[swap_id])?;
        let bytes = self
            .call_request(self.my_addr().await, swap_address, None, Some(data.into()))
            .await?;
        let decoded_tokens = function.decode_output(&bytes.0)?;

        let state = decoded_tokens.get(state_index).ok_or_else(|| {
```